include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -257,17 +258,6 @@ typedef enum OrtCudnnConvAlgoSearch {
   DEFAULT,     // default algorithm using CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_PRECOMP_GEMM
 } OrtCudnnConvAlgoSearch;
 
-/// <summary>
-/// Options for the CUDA provider that are passed to SessionOptionsAppendExecutionProvider_CUDA
-/// </summary>
-typedef struct OrtCUDAProviderOptions {
```

Comment:
Moved this struct to an internal header - this introduces only source incompatibility and ABI should still be preserved

Reply:
can we do this? what if a user is using SessionOptionsAppendExecutionProvider_CUDA() and directly using this struct?

---
In reply to: [553939121](https://github.com/microsoft/onnxruntime/pull/6291#discussion_r553939121) [](ancestors = 553939121)

Reply:
1) Can we do this ?
I believe we can. In their app, if people are contracting OrtCUDAProviderOptions and are calling  SessionOptionsAppendExecutionProvider_CUDA() by pasing in a pointer to the constructed instance, that app should still work with the latest version of ORT, wouldn't it (i.e.) ABI is preserved. They won't be able to re-build their app with the latest ORT source-code because the header is now missing the definition of `OrtCUDAProviderOptions` , so source compatibility doesn't exist - something we never guaranteed. They would have to use the API to create an instance of OrtCUDAProviderOptions to pass into SessionOptionsAppendExecutionProvider_CUDA(). We are trying to remove ambiguity in the way OrtCUDAProviderOptions can be constructed. We did something similar in https://github.com/microsoft/onnxruntime/pull/5634 for OrtArenaCfg. 
What do you think about this ?
CC: @pranavsharma

2) Can we move OrtCudnnConvAlgoSearch out as well ? Yes, I think so, moved it out.

Reply:
SessionOptionsAppendExecutionProvider_CUDA takes a pointer to OrtCUDAProviderOptions. The position and the signature of this function is not changing between v6 and v7. Hence in both old and new DLL cases, the correct function will be called. If the user is using the old header with the new DLL, she supplies the OrtCUDAProviderOptions ptr and all is good. New header with old DLL won't work with the same code (obviously) - unlikely scenario as there's no reason a user would upgrade without using the new DLL.